cmd/minishift/cmd/hostfolder/add.go

```diff
@@ -20,45 +20,223 @@ import (
+	addCmd.Flags().BoolVarP(&interactive, interactiveFlag, "i", false, "Allows to interactively provide the required parameters.")
+
+	// Windows-only
+	if runtime.GOOS == "windows" {
+		addCmd.Flags().BoolVar(&usersShare, usersShareFlag, false, "Defines the shared Users folder as the host folder on a Windows host.")
+	}
+}
+
+func addHostFolder(cmd *cobra.Command, args []string) {
+	hostFolderManager := getHostFolderManager()
+
+	var name string
+	if usersShare && runtime.GOOS == "windows" {
+		// Windows-only (CIFS), all instances
+		name = "Users"
+	} else {
+		if len(args) < 1 {
+			atexit.ExitWithMessage(1, usage)
 		}
+		name = args[0]
+	}
+
+	if hostFolderManager.Exist(name) {
+		atexit.ExitWithMessage(1, fmt.Sprintf("there is already a host folder with the name '%s' defined", name))
+	}
 
-		if err != nil {
-			atexit.ExitWithMessage(1, fmt.Sprintf("Failed to mount host folder: %s", err.Error()))
+	switch shareType {
+	case hf.CIFS.String():
+		if interactive {
+			addCIFSInteractive(hostFolderManager, name)
+		} else {
+			addCIFSNonInteractive(hostFolderManager, name)
+		}
+	case hf.SSHFS.String():
+		if interactive {
+			addSSHFSInteractive(hostFolderManager, name)
+		} else {
+			addSSHFSNonInteractive(hostFolderManager, name)
 		}
-	},
+	default:
+		atexit.ExitWithMessage(1, fmt.Sprintf(unknownType, shareType))
+	}
 }
 
-func init() {
-	HostfolderCmd.AddCommand(hostfolderAddCmd)
-	hostfolderAddCmd.Flags().BoolVarP(&instanceOnly, "instance-only", "", false, "Defines the host folder only for the current OpenShift cluster.")
+func addSSHFSInteractive(manager *hf.Manager, name string) error {
+	source := util.ReadInputFromStdin("source path")
+	mountPath := readInputForMountPoint(name)
 
-	// Windows-only
-	if runtime.GOOS == "windows" {
-		hostfolderAddCmd.Flags().BoolVarP(&usersShare, "users-share", "", false, "Defines the shared Users folder as the host folder on a Windows host.")
+	config := config.HostFolderConfig{
+		Name: name,
+		Type: hf.SSHFS.String(),
+		Options: map[string]string{
+			config.Source:     source,
+			config.MountPoint: mountPath,
+		},
+	}
+	hostFolder := hf.NewSSHFSHostFolder(config, miniConfig.AllInstancesConfig)
+	manager.Add(hostFolder, !instanceOnly)
+
+	return nil
+}
+
+func addSSHFSNonInteractive(manager *hf.Manager, name string) error {
+	if source == "" {
+		atexit.ExitWithMessage(1, noSource)
+	}
+
+	if target == "" {
+		atexit.ExitWithMessage(1, noTarget)
+	}
+
+	config := config.HostFolderConfig{
+		Name: name,
+		Type: hf.SSHFS.String(),
+		Options: map[string]string{
+			config.Source:     source,
+			config.MountPoint: target,
+		},
+	}
+	hostFolder := hf.NewSSHFSHostFolder(config, miniConfig.AllInstancesConfig)
+	manager.Add(hostFolder, !instanceOnly)
+
+	return nil
+}
+
+func addCIFSInteractive(manager *hf.Manager, name string) error {
+	var uncPath string
+	if usersShare {
+		uncPath = "[determined on startup]"
+	} else {
+		uncPath = util.ReadInputFromStdin("UNC path")
```

Comment:
Check for tty support is also relevant here. i would suggest to error out immediately on entry of this function when no tty support.. Or even earlier, like the switch for types

Reply:
Done